### PR TITLE
added GC to LLAPI

### PR DESCRIPTION
**Pull request by MeirShpilraien**

### CODE REVIEW

File: src/fork_gc.h [hunk 1]
```diff
@@ -18,11 +18,21 @@ typedef struct {
   uint64_t gcBlocksDenied;
 } ForkGCStats;
 
+typedef enum ForkGCCtxType {
```
Comment (mnunberg) [inline src/fork_gc.h]: If this enum isn't used anywhere except within forkgc; then this should just be in the .c file

File: src/gc.c [hunk 2]
```diff
@@ -45,7 +46,24 @@ static RedisModuleBlockedClient* BlockClients_pop(BlockClients* ctx){
...
 }
 
-GCContext* GCContext_CreateGC(RedisModuleString *keyName, float initialHZ, uint64_t uniqueId){
+GCContext* GCContext_CreateGCFromSpec(IndexSpec* sp, float initialHZ, uint64_t uniqueId,
+                                      uint32_t gcPolicy) {
+  GCContext* ret = rm_calloc(1, sizeof(GCContext));
+  pthread_mutex_init(&ret->bClients.lock, NULL);
+  switch (gcPolicy) {
+    case GCPolicy_Fork:
+      ret->gcCtx = (void*)NewForkGCFromSpec(sp, uniqueId, &ret->callbacks);
```
Comment (mnunberg) [inline src/gc.c]: why the cast?

File: src/redisearch_api.c [hunk 3]
```diff
@@ -30,7 +32,12 @@ static void valFreeCb(void* unused, void* p) {
 }
 
 static IndexSpec* RS_CreateIndex(const char* name, const RSIndexOptions* options) {
-  RSIndexOptions opts_s = {0};
+  RSIndexOptions opts_s = {
+      .gvcb = NULL,
```
Comment (mnunberg) [inline src/redisearch_api.c]: you don't need to initialize the other options; just the gcPolicy

File: src/redisearch_api.c [hunk 4]
```diff
@@ -107,6 +127,8 @@ static RSField* RS_CreateField(IndexSpec* sp, const char* name, unsigned types,
     sp->flags |= Index_HasPhonetic;
   }
 
+  RWLOCK_RELEASE();
```
Comment (mnunberg) [inline src/redisearch_api.c]: Missing a release on line 94. Use goto instead

File: src/redisearch_api.c [hunk 5]
```diff
@@ -204,6 +232,8 @@ static int RS_IndexAddDocument(IndexSpec* sp, Document* d, int options, char** e
   aCtx->stateFlags |= ACTX_F_NOBLOCK;
   AddDocumentCtx_Submit(aCtx, &sctx, options);
   rm_free(d);
+
+  RWLOCK_RELEASE();
```
Comment (mnunberg) [inline src/redisearch_api.c]: missing release on line 227; use goto

File: src/redisearch_api.c [hunk 6]
```diff
@@ -414,14 +447,18 @@ static void RS_ResultsIteratorFree(RS_ApiIter* iter) {
...
 }
 
 static void RS_ResultsIteratorReset(RS_ApiIter* iter) {
   iter->internal->Rewind(iter->internal->ctx);
 }
 
 static RSIndexOptions* RS_CreateIndexOptions() {
-  return rm_calloc(1, sizeof(RSIndexOptions));
+  RSIndexOptions* ret = rm_calloc(1, sizeof(RSIndexOptions));
+  ret->gcPolicy = GC_POLICY_NONE;
```
Comment (mnunberg) [inline src/redisearch_api.c]: Why not just make GC_POLICY_NONE the default at 0?

File: src/rwlock.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/*
```
Comment (mnunberg) [inline src/rwlock.h]: comment

File: src/spec.c [hunk 8]
```diff
@@ -726,13 +726,18 @@ void IndexSpec_Free(void *ctx) {
 }
 
 void IndexSpec_FreeSync(IndexSpec *spec) {
+  //  todo:
+  //  mark I think we only need IndexSpec_FreeInternals, this is called only from the
+  //  LLAPI and there is no need to drop keys cause its out of the key space.
+  //  Let me know what you think
+
```
Comment (mnunberg) [inline src/spec.c]: geo... I seem to recall that there was some keyspace interaction. Run the t_llapi test through valgrind, and see if indeed everything is actually deleted.

File: src/redisearch_api.c (hunk 4 above)
Comment (MeirShpilraien) [reply src/redisearch_api.c]: Nice catch

File: src/redisearch_api.c (hunk 5 above)
Comment (MeirShpilraien) [reply src/redisearch_api.c]: Nice catch :)

File: src/fork_gc.h (hunk 1 above)
Comment (MeirShpilraien) [reply src/fork_gc.h]: its been used by ForkGCCtx which is also defined on this h file is is used outside of it.

File: src/redisearch_api.c (hunk 3 above)
Comment (MeirShpilraien) [reply src/redisearch_api.c]: I think its cleaner this way

File: src/redisearch_api.c (hunk 6 above)
Comment (MeirShpilraien) [reply src/redisearch_api.c]: Because 0 is default GC

File: src/spec.c (hunk 8 above)
Comment (MeirShpilraien) [reply src/spec.c]: valgrind test failing with it or without it.
I will leave it with a todo to check this out.

File: src/redisearch_api.c (hunk 3 above)
Comment (mnunberg) [reply src/redisearch_api.c]: It's better to just 0-initialize it. This way you know the struct has its defaults; only overrides should be in the initializer.

**Comment by MeirShpilraien** [comment]
@mnunberg we need to think how we can test it

**Pull request merged**